keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+---
+title: Even Pods Spreading
+authors:
+  - "@Huang-Wei"
+owning-sig: sig-scheduling
+reviewers:
+  - "@wojtek-t"
+  - "@bsalamat"
+  - "@k82cn"
+approvers:
+  - "@bsalamat"
+  - "@k82cn"
+creation-date: 2019-02-21
+last-updated: 2019-03-11
+status: provisional
+---
+
+# Even Pods Spreading
+
+## Table of Contents
+
+* [Terms](#terms)
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+    * [Story 1 - NodeAffinity](#story-1---nodeaffinity)
+    * [Story 2 - PodAffinity](#story-2---podaffinity)
+    * [Story 3 - PodAntiAffinity](#story-3---podantiaffinity)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Design Details](#design-details)
+  * [Design A](#design-a)
+  * [Design B](#design-b)
+  * [Algorithm](#algorithm)
+    * [NodeAffinity](#nodeaffinity)
+    * [PodAffinity](#podaffinity)
+    * [PodAntiAffinity](#podantiaffinity)
+  * [Pros/Cons](#proscons)
+  * [Conclusion](#conclusion)
+  * [Test Plan](#test-plan)
+  * [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Terms
+
+- **Topology:** describe a series of worker nodes which belongs to the same
+  region/zone/rack/hostname/etc. In terms of Kubernetes, they're defined and
+  grouped by node labels.
+- **Affinity**: if not specified particularly, "Affinity" refers to
+  `NodeAffinity`, `PodAffinity` and `PodAntiAffinity`.
+- **CA**: Cluster Autoscaler.
+  [CA](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
+  is a tool that automatically adjusts the size of the Kubernetes cluster upon
+  specific conditions.
+
+## Summary
+
+`EvenPodsSpreading` feature gives users more fine-grained control on
+distribution of pods scheduling, so as to achieve better high availability and
+resource utilization.
+
+## Motivation
+
+In Kubernetes, "Affinity" related directives are aimed to control how pods are
+scheduled - more packed or more scattering. But right now only limited options
+are offered: for `PodAffinity`, infinite pods can be stacked onto qualifying
+topology domain(s); for `PodAntiAffinity`, pods are scheduled onto each single
+topology domain exclusively.
```

Comment:
what does single domain mean ? Each pod anti affinity describes the domain and spreading happens accordingly. Do you want the spreading to take into account multiple topology domains simultaneously ?

Reply:
I'm not sure I followed your question. I meant for `PodAntiAffinity`, right now we can't put more than one pod onto the same topology domain.

Reply:
thanks, that helps. Isnt that true for both pod affinity and anti affinity ?

Reply:
Not exactly. For pod affinity, it's possible (although rare) incoming pods can be scheduled onto multiple topology domains. For example, we have a 3-zone cluster, and each zone has pods with label `foo`. Now a pod comes in, its PodAffinityTerm desires to be scheduled to a zone with pod having label `foo`, in this case, any zone can be a fit.

Essentially speaking, it's b/c the podAffinityTerm API doesn't enforce you to match with yourself, it can be any conbination form of selector+namespace+topologyKey.